jetty-webapp/src/main/java/org/eclipse/jetty/webapp/DecoratingListener.java

```diff
@@ -0,0 +1,180 @@
+//
+//  ========================================================================
+//  Copyright (c) 1995-2019 Mort Bay Consulting Pty. Ltd.
+//  ------------------------------------------------------------------------
+//  All rights reserved. This program and the accompanying materials
+//  are made available under the terms of the Eclipse Public License v1.0
+//  and Apache License v2.0 which accompanies this distribution.
+//
+//      The Eclipse Public License is available at
+//      http://www.eclipse.org/legal/epl-v10.html
+//
+//      The Apache License v2.0 is available at
+//      http://www.opensource.org/licenses/apache2.0.php
+//
+//  You may elect to redistribute this code under either of these licenses.
+//  ========================================================================
+//
+
+package org.eclipse.jetty.webapp;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import javax.servlet.ServletContextAttributeEvent;
+import javax.servlet.ServletContextAttributeListener;
+
+import org.eclipse.jetty.util.Decorator;
+import org.eclipse.jetty.util.log.Log;
+import org.eclipse.jetty.util.log.Logger;
+
+/**
+ * A ServletContextAttributeListener that listens for a specific attribute
+ * name (default "org.eclipse.jetty.webapp.Decorator") to obtain a
+ * decorator instance from the webapp.  The instance is then either coerced
+ * to a Decorator or reflected for decorator compatible methods so it can
+ * be added to the {@link WebAppContext#getObjectFactory()} as a
+ * {@link Decorator}.
+ */
+public class DecoratingListener implements ServletContextAttributeListener
+{
+    public static final String DECORATOR_ATTRIBUTE = "org.eclipse.jetty.webapp.decorator";
+    private static final Logger LOG = Log.getLogger(DecoratingListener.class);
+    private static final MethodType decorateType;
+    private static final MethodType destroyType;
+
+    static
+    {
+        try
+        {
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            decorateType = MethodType.methodType(Object.class, Object.class);
+            destroyType = MethodType.methodType(Void.TYPE, Object.class);
+            // Ensure we have a match
+            lookup.findVirtual(Decorator.class, "decorate", decorateType);
+            lookup.findVirtual(Decorator.class, "destroy", destroyType);
```

Comment:
I don't understand the 2 lines above. They are looking up a method of that signature in the `Decorator` class, which does define those 2 methods, so is this check a safety net if the `Decorator` class gets changes?
Otherwise looks unnecessary to me (it's like testing if the `Runnable` class has a `void run()` method).

Reply:
yes - as I said last time.... it's a safety net.